subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -330,146 +320,26 @@ int nrf_cloud_coap_connect(const char * const app_ver)
-			MODEM_INFO_FWVER_SIZE +
-			strlen(BUILD_VERSION_STR) +
-			strlen(CDDL_VERSION)];
-
-	err = modem_info_get_fw_version(mfw_string, sizeof(mfw_string));
-	if (!err) {
-		err = snprintf(ver_string, sizeof(ver_string), VER_STRING_FMT,
-			       mfw_string, BUILD_VERSION_STR, CDDL_VERSION);
-		if ((err < 0) || (err >= sizeof(ver_string))) {
-			LOG_ERR("Could not format string");
-			return -ETXTBSY;
-		}
-	} else {
-		LOG_ERR("Unable to obtain the modem firmware version: %d", err);
-	}
-#else
-	char *ver_string = VER_STRING_FMT2;
-#endif
-
-	LOG_DBG("Generate JWT");
-	jwt = nrf_cloud_malloc(JWT_BUF_SZ);
-	if (!jwt) {
-		return -ENOMEM;
-	}
-	err = nrf_cloud_jwt_generate(NRF_CLOUD_JWT_VALID_TIME_S_MAX, jwt, JWT_BUF_SZ);
-	if (err) {
-		LOG_ERR("Error generating JWT with modem: %d", err);
-		nrf_cloud_free(jwt);
-		return err;
-	}
-
-	LOG_INF("Request authorization with JWT");
-	err = nrf_cloud_coap_post("auth/jwt", err ? NULL : ver_string,
-				 (uint8_t *)jwt, strlen(jwt),
-				 COAP_CONTENT_FORMAT_TEXT_PLAIN, true, auth_cb, NULL);
-	nrf_cloud_free(jwt);
-
-	if (err) {
-		return err;
-	}
-	if (!authenticated) {
-		return -EACCES;
-	}
-
-	LOG_INF("Authorized");
-
-	if (nrfc_dtls_cid_is_active(sock)) {
-		LOG_INF("DTLS CID is active");
-	}
+	nrf_cloud_coap_transport_disconnect(&internal_cc);
 	return err;
 }
 
 int nrf_cloud_coap_pause(void)
 {
-	int err = 0;
-
-	cid_saved = false;
-
-	if (nrfc_dtls_cid_is_active(sock) && authenticated) {
-		err = nrfc_dtls_session_save(sock);
-		if (!err) {
-			LOG_DBG("DTLS CID session saved.");
-			cid_saved = true;
-		} else {
-			LOG_ERR("Unable to save DTLS CID session: %d", err);
-		}
-	} else {
-		LOG_WRN("Unable to pause CoAP connection.");
-		err = -EACCES;
-	}
-	return err;
+	return nrf_cloud_coap_transport_pause(&internal_cc);
 }
 
 int nrf_cloud_coap_resume(void)
 {
-	int err = 0;
-
-	if (cid_saved) {
-		err = nrfc_dtls_session_load(sock);
-		cid_saved = false;
-		if (!err) {
-			LOG_DBG("Loaded DTLS CID session");
-			authenticated = true;
-		} else if (err == -EAGAIN) {
-			LOG_INF("Failed to load DTLS CID session");
-		} else if (err == -EINVAL) {
-			LOG_INF("DLTS CID sessions not supported with current modem firmware");
-		} else {
-			LOG_ERR("Error on DTLS CID session load: %d", err);
-		}
-	} else {
-		LOG_WRN("Unable to resume CoAP connection.");
-		err = -EACCES; /* Cannot resume. */
-	}
-
-	return err;
+	return nrf_cloud_coap_transport_resume(&internal_cc);
 }
 
-static K_SEM_DEFINE(serial_sem, 1, 1);
-static K_SEM_DEFINE(cb_sem, 0, 1);
-
-struct user_cb {
-	coap_client_response_cb_t cb;
-	void *user_data;
-	int result_code;
-};
-
 static void client_callback(int16_t result_code, size_t offset, const uint8_t *payload, size_t len,
 			    bool last_block, void *user_data)
 {
-	struct user_cb *user_cb = (struct user_cb *)user_data;
+	struct cc_xfer_data *xfer = (struct cc_xfer_data *)user_data;
 
-	user_cb->result_code = result_code;
+	xfer->result_code = result_code;
```

Comment:
Below you check whether xfer is NULL before using it. You should do that here too.

Reply:
xfer must always be valid... will replace with `__ASSERT_NO_MSG(user_data != NULL);`